### PR TITLE
Disallow cascade_via_foreign_keys if any partition rel has non-inherited fkeys

### PR DESCRIPTION
**Pull request by onurctirtir**
Fixes the issue described in the first comment of https://github.com/citusdata/citus/issues/4481.

### CODE REVIEW

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 1]
```diff
@@ -116,6 +143,43 @@ LockRelationsWithLockMode(List *relationIdList, LOCKMODE lockMode)
 }
 
 
+/*
+ * HasPartitionRelationInvolvedInAnyNonInheritedForeignKeys is a wrapper function that
+ * returns true if FindPartitionRelationInvolvedInAnyNonInheritedForeignKeys can find
+ * a matching relation.
+ */
+static bool
+HasPartitionRelationInvolvedInAnyNonInheritedForeignKeys(List *relationIdList)
```
Comment (onurctirtir) [inline src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:152]: thinking about better names for `HasPartitionRelationInvolvedInAnyNonInheritedForeignKeys`, `FindPartitionRelationInvolvedInAnyNonInheritedForeignKeys` & `RelationInvolvedInAnyNonInheritedForeignKeys`

Comment (onurctirtir) [reply src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c]: removed those functions, but finding a better name still applies to `ErrorIfHasPartitionRelationInvolvedInNonInheritedForeignKey`

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 2]
```diff
@@ -68,6 +70,16 @@ CascadeOperationForConnectedRelations(Oid relationId, LOCKMODE lockMode,
...
 	LockRelationsWithLockMode(fKeyConnectedRelationIdList, lockMode);
 
+	/*
+	 * Before removing any partition relations, we should error out here if any
+	 * of connected relations is a partition table involved in a foreign key that
+	 * is not inherited from its parent table.
+	 * We should handle this case here as we remove partition relations in this
+	 * function	before ExecuteCascadeOperationForRelationIdList.
+	 */
+	ErrorIfHasPartitionRelationInvolvedInNonInheritedForeignKey(
```
Comment (onderkalaci) [inline src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:80]: `ErrorIfHas` -> `ErrorIfAny`

File: src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c [hunk 3]
```diff
@@ -116,6 +128,41 @@ LockRelationsWithLockMode(List *relationIdList, LOCKMODE lockMode)
...
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
+	{
+		if (!PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		Oid partitionRelationId = relationId;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/cascade_table_operation_for_connected_relations.c:148]: Minor: I think this extra variable seems redundant

File: src/test/regress/expected/undistribute_table_cascade.out [hunk 4]
```diff
@@ -308,6 +308,28 @@ BEGIN;
...
 (6 rows)
 
+ROLLBACK;
+BEGIN;
+  ALTER TABLE partitioned_table_1_100_200 ADD CONSTRAINT non_inherited_fkey FOREIGN KEY(col_1) REFERENCES reference_table_3(col_1);
+  SELECT undistribute_table('partitioned_table_2', cascade_via_foreign_keys=>true);
+ERROR:  cannot cascade operation via foreign keys as partition table undistribute_table_cascade.partitioned_table_1_100_200 involved in a foreign key relationship that is not inherited from it's parent table
+ROLLBACK;
+BEGIN;
+  ALTER TABLE partitioned_table_1_100_200 ADD CONSTRAINT non_inherited_fkey FOREIGN KEY(col_1) REFERENCES partitioned_table_2_100_200(col_1);
```
Comment (onderkalaci) [inline src/test/regress/expected/undistribute_table_cascade.out:318]: can you add one test with self referencing partition?

**Comment by onurctirtir** [comment]
@onderkalaci found, this is still not covered by this pr:

```sql
CREATE TABLE partitioned_table_1 (col_1 INT UNIQUE, col_2 INT) PARTITION BY RANGE (col_1);
CREATE TABLE partitioned_table_1_100_200 PARTITION OF partitioned_table_1 FOR VALUES FROM (100) TO (200);
CREATE TABLE partitioned_table_1_200_300 PARTITION OF partitioned_table_1 FOR VALUES FROM (200) TO (300);
SELECT create_distributed_table('partitioned_table_1', 'col_1');


CREATE TABLE distributed_table_1 (col_1 INT UNIQUE);
SELECT create_distributed_table('distributed_table_1', 'col_1');
ALTER TABLE distributed_table_1 ADD CONSTRAINT fkey FOREIGN KEY(col_1) REFERENCES partitioned_table_1_100_200(col_1);
SELECT undistribute_table('partitioned_table_1', true);
ERROR:  cannot undistribute table because a foreign key references to it
```

**Comment by onderkalaci** [comment]
> @onderkalaci found, this is still not covered by this pr:


this is covered now?

**Comment by onurctirtir** [comment]
> > @onderkalaci found, this is still not covered by this pr:
> 
> this is covered now?

yes https://github.com/citusdata/citus/pull/4487/files#diff-98dd7367bf6e16a6c8fb9a1d404d9852bd205be8861e1e336032aaaf11e35c33R292-R297

**Review by onderkalaci — APPROVED** [review]
Let me summarize our in-person chat here. 

This limitation may not be very intuitive for the users. Still, we don't think it is a common pattern to have a foreign key from/to a partition instead of the parent table. Also, given that the workaround is to just drop the foreign key, we decided to go ahead and merge it for now.

If this becomes a friction point in the future, we can change the code-flow a bit and fix it properly.

**Pull request merged**